crypto/asn1_dsa.c

```diff
@@ -0,0 +1,298 @@
+            break;
+        }
+        *out++ = (unsigned char)cont_len;
+        *ppout = out;
+    }
+    return encoded_len;
+}
+
+/*
+ * Outputs the DER encoding of a positive ASN.1 INTEGER to *ppout and, if
+ * successful, increments *ppout past the data just written.
+ * If n is negative then an error results.
+ * If ppout is NULL then the output size is calculated and returned but no
+ * output is produced.
+ * An attempt to produce more than len bytes results in an error.
+ * Returns the number of bytes of output produced or 0 if an error occurs.
+ */
+size_t encode_der_integer(const BIGNUM *n, unsigned char **ppout, size_t len)
+{
+    unsigned char *out = NULL;
+    unsigned char **pp = NULL;
+    size_t produced;
+    size_t c;
+    size_t cont_len;
+
+    if (len < 1 || BN_is_negative(n))
+        return 0;
+
+    /*
+     * Calculate the ASN.1 INTEGER DER content length for n.
+     * This is the number whole bytes required to represent n (i.e. rounded
+     * down), plus one.
+     * If n is zero then the content is a single zero byte (length = 1).
+     * If the number of bits of n is a multiple of 8 then an extra zero padding
+     * byte is included to ensure that the value is still treated as positive
+     * in the INTEGER two's complement representation.
+     */
+    cont_len = BN_num_bits(n) / 8 + 1;
+
+    if (ppout != NULL) {
+        out = *ppout;
+        pp = &out;
+        *out++ = ID_INTEGER;
+    }
+    produced = 1;
+    if ((c = encode_der_length(cont_len, pp, len - produced)) == 0)
+        return 0;
+    produced += c;
+    if (cont_len > len - produced)
+        return 0;
+    if (pp != NULL) {
+        if (BN_bn2binpad(n, out, (int)cont_len) != (int)cont_len)
+            return 0;
+        out += cont_len;
+        *ppout = out;
+    }
+    produced += cont_len;
+    return produced;
+}
+
+/*
+ * Outputs the DER encoding of a DSA-Sig-Value or ECDSA-Sig-Value to *ppout
+ * and increments *ppout past the data just written.
+ * If ppout is NULL then the output size is calculated and returned but no
+ * output is produced.
+ * An attempt to produce more than len bytes results in an error.
+ * Returns the number of bytes of output produced (or that would be produced)
+ * or 0 if an error occurs.
+ */
+size_t encode_der_dsa_sig(const BIGNUM *r, const BIGNUM *s,
+                          unsigned char **ppout, size_t len)
+{
+    unsigned char *out = NULL;
+    unsigned char **pp = NULL;
+    size_t produced;
+    size_t c;
+    size_t r_der_len;
+    size_t s_der_len;
+    size_t cont_len;
+
+    if (len < 1
+            || (r_der_len = encode_der_integer(r, NULL, SIZE_MAX)) == 0
+            || (s_der_len = encode_der_integer(s, NULL, SIZE_MAX)) == 0)
+        return 0;
+
+    cont_len = r_der_len + s_der_len;
+
+    if (ppout != NULL) {
+        out = *ppout;
+        pp = &out;
+        *out++ = ID_SEQUENCE;
+    }
+    produced = 1;
+    if ((c = encode_der_length(cont_len, pp, len - produced)) == 0)
```

Comment:
what happens in len was passed in as 2 here?

Reply:
In every encode function, and at each step, `produced` is never greater than `len` so `len - produced` can never be negative.
At the line you commented on, if `len` was passed in as `2` and `produced` is `1` (from the line above, `produced = 1;`) then `len - produced` is `2 - 1`, which is `1` (which is positive).

Reply:
yep ok..  so that is why the initial len < 1 check is needed. I would normally decrement the total length to do this. But your way also works.

Reply:
Yeah, I could have been more verbose: initialized `produced` to `0`, checked if `1 > len - produced` before emitting `ID_SEQUENCE`, then incremented `produced`, etc., but I thought that was unnecessary.
What do you think?

Reply:
no its ok

Reply:
> yep ok.. so that is why the initial len < 1 check is needed. I would normally decrement the total length to do this. But your way also works.

Sure, reducing the length at each step would have worked. The return value (bytes produced) could have been calculated with something like `orig_len - len` or by pointer difference and saving the original out pointer.